src/operator/tensor/broadcast_reduce_op.h

```diff
@@ -880,17 +878,30 @@ inline bool L2NormStorageType(const nnvm::NodeAttrs& attrs,
   CHECK_EQ(out_attrs->size(), 1U);
   const int in_stype = in_attrs->at(0);
   int& out_stype = out_attrs->at(0);
+  const NormParam& param = nnvm::get<NormParam>(attrs.parsed);
   bool dispatched = false;
+  // l2 norm on a particular axis only supports cpu
+  const bool invalid_ctx = dev_mask != mshadow::cpu::kDevMask;
+  const auto dispatch_ex =
+      invalid_ctx ? DispatchMode::kFComputeFallback : DispatchMode::kFComputeEx;
   if (!dispatched && in_stype == kDefaultStorage) {
     // dns -> dns
     dispatched = storage_type_assign(&out_stype, kDefaultStorage, dispatch_mode,
                                      DispatchMode::kFCompute);
   }
-  if (!dispatched && (in_stype == kCSRStorage || in_stype == kRowSparseStorage)) {
-    // csr/rsp -> dns
+  const TShape axis = param.axis.has_value() ? param.axis.value() : TShape();
+  if (!dispatched && (in_stype == kRowSparseStorage || in_stype == kCSRStorage) &&
+      axis.ndim() == 0 && !param.keepdims && param.ord == 2) {
```

Comment:
does it matter whether keepdims is true?

Reply:
Good catch. I should probably remove this. Actually I found that when axis is None, keepdims has no effect for the dense implementation. 
```

>>> import mxnet as mx
>>> a = mx.nd.ones((2,2))
>>> a.norm()

[ 2.]
<NDArray 1 @cpu(0)>
>>> a.norm(keepdims=True)

[ 2.]
<NDArray 1 @cpu(0)>
```
Is this supposed to return a result of shape (1,1) instead?